metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala

```diff
@@ -44,47 +110,375 @@ class PackageProvider(private val buildTargets: BuildTargets) {
+    tree.collect { case p: Pkg => p }.toVector
+
+  private def wrap(str: String): String = Identifier.backtickWrap(str)
 
-  private def workspaceEdit(
+  private def mkWorkspaceEdit(
       path: AbsolutePath,
-      packageStatement: String,
+      replacement: String,
+      range: Range = new Range(new Position(0, 0), new Position(0, 0)),
   ): WorkspaceEdit = {
-    val textEdit = new TextEdit(
-      new Range(new Position(0, 0), new Position(0, 0)),
-      packageStatement,
-    )
+    val textEdit = new TextEdit(range, replacement)
     val textEdits = List(textEdit).asJava
     val changes = Map(path.toURI.toString -> textEdits).asJava
     new WorkspaceEdit(changes)
   }
+
+  private def deducePkgParts(path: AbsolutePath): Vector[String] =
+    calcPathToSourceRoot(path).dropRight(1)
+
+  private def calcPathToSourceRoot(
+      path: AbsolutePath
+  ): Vector[String] =
+    buildTargets
+      .inverseSourceItem(path)
+      .map(path.toRelative)
+      .flatMap(relativePath => Option(relativePath.toNIO))
+      .map { _.iterator().asScala.map(p => wrap(p.toString())).toVector }
+      .getOrElse(Vector.empty)
+
+  /**
+   * Splits package declaration into statements
+   * Desired property - not to break old visibility regions
+   * @param oldPackageSplit: content of package statements before refactoring
+   * @param newPkgs: list of package parts that should be present after refactoring
+   * @return split of `newPkgs` into package statements with respect to the `oldPackageSplit`
+   */
+  private def splitPackageStatements(
+      oldPackageSplit: Vector[Vector[String]],
+      newPkgs: Vector[String],
+  ): Vector[Vector[String]] = {
+    val oldPackageSplitSet = oldPackageSplit.toSet
+    val oldEndPkgs =
+      oldPackageSplit.flatMap(_.lastOption).reverse.zipWithIndex.toMap
+
+    var consideredEndPkg = 0
+    var pkgStmts: Vector[Vector[String]] = Vector.empty
+    var pkgStmt: Vector[String] = Vector.empty
+
+    newPkgs.reverse.foreach { newPkg =>
+      val foundEndPkg =
+        oldEndPkgs.get(newPkg).filter { _ >= consideredEndPkg }
+      foundEndPkg match {
+        case Some(endPkgInd) => {
+          if (endPkgInd == 0) pkgStmt = newPkg +: pkgStmt
+          else {
+            if (pkgStmt.nonEmpty) pkgStmts = pkgStmt +: pkgStmts
+            pkgStmt = Vector(newPkg)
+          }
+          consideredEndPkg = endPkgInd + 1
+        }
+        case None =>
+          if (oldPackageSplitSet.contains(pkgStmt)) {
+            pkgStmts = pkgStmt +: pkgStmts
+            pkgStmt = Vector.empty
+          }
+          pkgStmt = newPkg +: pkgStmt
+      }
+    }
+    if (pkgStmt.nonEmpty) pkgStmts = pkgStmt +: pkgStmts
+    pkgStmts
+  }
+
+  @tailrec
+  private def extractNames(
+      t: Term,
+      acc: Vector[String] = Vector.empty,
+  ): Vector[String] = {
+    t match {
+      case n: Term.Name => n.value +: acc
+      case s: Term.Select => extractNames(s.qual, s.name.value +: acc)
+    }
```

Comment:
Not sure if Vector is needed here. As I see here you append at the beginning

Reply:
Imho, Vector since scala 2.13 is a good default collection as it provides quite efficient prepend, append, concat, etc. operations. Replacing returning type from vector to list here would proliferate with `Vector[List[..]]` types in other places, which looks somewhat inconsistent 🤔

Reply:
yep, it's good one but it's anyway it's needed in case if you use append and random access.
Maybe I missed smth in PR but I haven't noticed such usage and looks like List should fit well here and `List[List[String]]` instead of vector in other places

Reply:
There are  `lastOption`  ([one](https://github.com/susliko/metals/blob/cf937e8f01a4c4f8e6baa32d53773aafd9f57a59/metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala#L372), [two](https://github.com/susliko/metals/blob/cf937e8f01a4c4f8e6baa32d53773aafd9f57a59/metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala#L428), etc.) and `size` ([one](https://github.com/susliko/metals/blob/cf937e8f01a4c4f8e6baa32d53773aafd9f57a59/metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala#L231)) calls